wire-library/wire-runtime/src/darwinMain/kotlin/com/squareup/wire/ProtoAdapter.kt

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2020 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire
+
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.usePinned
+import platform.Foundation.NSData
+import platform.posix.memcpy
+
+fun <E> ProtoAdapter<E>.decode(data: NSData): E {
+  val bytes = ByteArray(data.length.toInt()).apply {
```

Comment:
can we `require` the length be in bounds first rather than silently truncating?

Reply:
not sure if you can test that... does a 2.0000000001GiB empty NSData actually allocate?

Reply:
By truncating you mean going from unsigned to signed? There's [`dataWithCapacity`](https://developer.apple.com/documentation/foundation/nsmutabledata/1547236-datawithcapacity?language=objc) which doesn't allocate, but it looks like it sets `length` to 0.

Reply:
Yes. The `toInt()` to me indicates `NSData` can be larger than `ByteArray` so I want to make sure we throw in that case.

Reply:
Added the require, no tests cause there doesn't seem to be a way to create a large enough `NSData` without allocating a ton of memory. Also added the `@ExperimentalUnsignedTypes` annotation to the method since `NSData`'s `length` is an `NSUInteger`, which Kotlin converts to `ULong`. Does this work, or do we want to introduce our own opt in marker?

Reply:
This seems fine. Perhaps note it in the method docs (which should probably be added regardless).

Reply:
Yep, added the docs, will merge on green.